src/iop/basecurve.c

```diff
@@ -1213,54 +1313,26 @@ void process_lut(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, co
 {
   const float *const in = (const float *)ivoid;
   float *const out = (float *)ovoid;
-  const int ch = piece->colors;
+  //const int ch = piece->colors; <-- it appears someone was trying to make this handle monochrome data, however the for loops only handled RGBA
```

Comment:
This looks wrong to me. The number of colors per channel should not be encoded anywhere. IIRC this value is 4 most of the time but we don't want to assume this in the code anywhere.

Reply:
I think we found a flaw in the existing code.  If you read lower down in the function, you'll see, in the legacy curve function:
`for(int i = 0; i < 3; i++)` - if you read what's in that loop, followed by the `outp[3] = inp[3]`, it's performing per-channel lookup on R, G, and B, then passing through what I assume is intended to be A.

Again, in the new apply_curve, we have:
```
    for(size_t c = 0; c < 3; c++)
    {
      outp[c] = (ratio * inp[c]);
    }
    outp[3] = inp[3];
```

Again, - pretty clearly written for RGBA.

I think you've found an existing bug regarding failure to handle the monochrome case.  After all, "preserve colors" really has no meaning with monochrome data - so fixing those issues probably needs a bunch of additional work (either by adding an additional commit to this pull request, or I promise to work on that next in a subsequent pull request before pushing up my exposure fusion changes).

Although in the discussion of https://github.com/darktable-org/darktable/pull/2787 - it sounds like at some point in the pipeline, monochrome images are converted into RGB (with equal values for each of R, G, and B) - under what conditions would we see a channel count other than 4 at this point in the pipe?  What would the pixel format be in this case?